### ISSUE TITLE
MIDI port connection changes not detected

### ISSUE THREAD

**Issue by Erriez** [pygame/pygame#3301]
**Environment:**

- **Operating system**: Windows 10
- **Python version** : 3.91.0
- **SDL version**: N.A.
- **PyGame version** : 2.1.2
- **Relevant hardware**: N.A.

**Current behavior:**

The function `pygame.midi.get_count()` always returns the same number of MIDI ports after startup, even after a MIDI device connect/disconnect. 

Application background: When clicking on a drop-down box, it should update new connected/disconnected MIDI ports.

**Expected behavior:**

Update number of MIDI ports after connect/disconnect.

**Screenshots**

N.A.

**Test code**

1. Run test code:

```python
import pygame

pygame.init()
pygame.midi.init()

while True:
    print(pygame.midi.get_count())
    time.sleep(1)
```

Output for example: 2 MIDI ports.

2. Connect MIDI device.
3. Output is still 2 MIDI ports.
4. Restart Python script
5. Output is for example: 4 MIDI ports.

**Stack trace/error output/other error logs**

N.A.

**Comment by Starbuck5** [comment]
Would it work if you quit and re init the midi module?

```
pygame.midi.quit()
pygame.midi.init()
```

I don't think the portmidi backend has hotplugging support, see: https://github.com/mixxxdj/portmidi/issues/8#issuecomment-933869739

**Comment by Erriez** [comment]
I created a test case which calls the `quit()` and `init()` functions before `get_count()` and then the number of ports is updated correctly. However, open MIDI connections are closed which is a problem for my application design.

Testcase:

```python
import pygame
import pygame.midi
import time


def get_mini_ports():
    # Hack to re-initialize MIDI to return updated `get_count()`
    # Open MIDI connections are closed!
    pygame.midi.quit()
    pygame.midi.init()

    midi_ports = []
    for i in range(pygame.midi.get_count()):
        (midi_interface, midi_name, midi_input, midi_output, opened) = pygame.midi.get_device_info(i)

        midi_interface = midi_interface.decode('utf-8')
        midi_name = midi_name.decode('utf-8')

        midi_ports.append({
            'ID': i,
            # 'midi_interface': midi_interface,
            'midi_name': midi_name,
            'midi_input': midi_input,
            'midi_output': midi_output,
            'opened': opened
        })
    return midi_ports


def print_mini_ports(portout):
    # Send some Sysex data
    portout.write_sys_ex(pygame.midi.time(), [0xf0, 0x00, 0x00, 0x00, 0x00, 0xf7])
    print('Write before MIDI quit()/init(): OK')

    print('MIDI ports: ')
    for port in get_mini_ports():
        print('  {}'.format(port))

    # Send some Sysex data
    portout.write_sys_ex(pygame.midi.time(), [0xf0, 0x00, 0x00, 0x00, 0x00, 0xf7])
    print('Write after MIDI quit()/init(): OK')


if __name__ == '__main__':
    # Initialize Pygame
    pygame.init()
    pygame.fastevent.init()
    pygame.midi.init()

    # Open MIDI port
    portout = pygame.midi.Output(1)

    while True:
        print_mini_ports(portout)
        time.sleep(1)
```

> I don't think the portmidi backend has hotplugging support, see: https://github.com/mixxxdj/portmidi/issues/8#issuecomment-933869739

Thanks for your confirmation.

It looks like correctly handled by Python https://github.com/mido/mido package, but need to investigate further:

```python
import mido
import time

# Connect/disconnect MIDI devices while printing MIDI ports:
while True:
    print('MIDI inputs:')
    print(mido.get_input_names())
    print('MIDI outputs:')
    print(mido.get_output_names())
    time.sleep(1)
```

**Comment by Starbuck5** [comment]
> It looks like correctly handled by Python https://github.com/mido/mido package, but need to investigate further:

See, it's funny you say that, since mido says it has three backends:
- pygame
- portmidi (pygame's backend)
- rtmidi (unrelated)

**Comment by Erriez** [comment]
You're right, I used the mido API with backend `python-rtmidi`. After some additional investigation, I found a bug in `python-rtmidi`: Incorrect MIDI port names. I reported the bug in https://github.com/mido/mido/issues/381, https://github.com/SpotlightKid/python-rtmidi/issues/111.

So I'll continue with Pygame and implement a workaround where the MIDI ports are not updated after init and connect/disconnect. In this case the user should always disconnect an open MIDI port, then the MIDI ports are updated correctly as you suggested.
